backend/api/src/websockets/server.ts

```diff
@@ -0,0 +1,119 @@
+import { Server as HttpServer } from 'node:http'
+import { Server as WebSocketServer, RawData, WebSocket } from 'ws'
+import { isError } from 'lodash'
+import { log } from 'shared/utils'
+import { Switchboard } from './switchboard'
+import {
+  ClientMessage,
+  ServerMessage,
+  CLIENT_MESSAGE_SCHEMA,
+} from 'common/api/websockets'
+
+const SWITCHBOARD = new Switchboard()
+
+export class MessageParseError extends Error {
+  details?: unknown
+  constructor(message: string, details?: unknown) {
+    super(message)
+    this.name = 'MessageParseError'
+    this.details = details
+  }
+}
+
+function serializeError(err: unknown) {
+  return isError(err) ? err.message : 'Unexpected error.'
+}
+
+function parseMessage(data: RawData): ClientMessage {
+  let messageObj: any
+  try {
+    messageObj = JSON.parse(data.toString())
+  } catch (err) {
+    log.error(err)
+    throw new MessageParseError('Message was not valid UTF-8 encoded JSON.')
+  }
+  const result = CLIENT_MESSAGE_SCHEMA.safeParse(messageObj)
+  if (!result.success) {
+    const issues = result.error.issues.map((i) => {
+      return {
+        field: i.path.join('.') || null,
+        error: i.message,
+      }
+    })
+    throw new MessageParseError('Error parsing message.', issues)
+  } else {
+    return result.data
+  }
+}
+
+function processMessage(ws: WebSocket, data: RawData): ServerMessage<'ack'> {
+  try {
+    const msg = parseMessage(data)
+    const { type, txid } = msg
+    try {
+      switch (type) {
+        case 'identify': {
+          SWITCHBOARD.identify(ws, msg.uid)
+          break
+        }
+        case 'subscribe': {
+          SWITCHBOARD.subscribe(ws, ...msg.topics)
+          break
+        }
+        case 'unsubscribe': {
+          SWITCHBOARD.unsubscribe(ws, ...msg.topics)
+          break
+        }
+        case 'ping': {
+          SWITCHBOARD.markSeen(ws)
+          break
+        }
+        default:
+          throw new Error("Unknown message type; shouldn't be possible here.")
+      }
+    } catch (err) {
+      log.error(err)
+      return { type: 'ack', txid, success: false, error: serializeError(err) }
+    }
+    return { type: 'ack', txid, success: true }
+  } catch (err) {
+    log.error(err)
+    return { type: 'ack', success: false, error: serializeError(err) }
+  }
+}
+
+export function broadcast(topic: string, data: Record<string, unknown>) {
```

Comment:
Would we want some kind of event type system? I think supabase channels worked this way with multiple event types per channel/topic.

Probably can just add it later though if we need it.

Reply:
We'll see if we need it. FYI, in the Supabase implementation the events are just a client abstraction anyway. You couldn't actually ask for only some events from the server. The topics were the real thing.

Reply:
Sounds good!